Question suggestions UI

nithusha21 opened the pull request:
## Explanation
This PR adds the basic questions suggestions UI. It adds a create question button on the creator dashboard, adds another table in the topics and skills dashboard which displays a list of suggested questions and their statuses. On clicking the row, it shows a view which displays the full question, and the reviewer can either accept, or reject with some review message.

This PR is WIP because there are some bugs which block this PR. Along with @aks681's help we'll try to fix these issues. Known bugs:
- Unable to add a solution from the create question modal in the creator dashboard.
- On using an existing question state data to populate a question editor, the view is completely broken

For manual testing, the following flags need to be enabled:
- constants.USE_NEW_SUGGESTIONS_FRAMEWORK
- feconf.ENABLE_GENERALISED_FEEDBACK_THREADS
- feconf.ENABLE_NEW_STRUCTURES

After that, give yourself admin role, in order to view the topics and skills dashboard, accessible at /topics_and_skills_dashboard
Then do the following
- Create a topic
- Create a skill, and add the skill to the topic's list of skills
- Now create a question, choose the topic from the dropdown
- In the questions tab of the topics_and_skills dashboard, the question will be visible
- Click on it to view the question editor (It is broken right now, but hopefully it will be fixed!)
- Click on select skill button, to view a list of skills in the topic which you can attach the question to
- Add a review message in the text box (Optional if accepting)
- Accept/reject the suggestion

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [ ] The PR explanation includes the words "Fixes #bugnum: ...".
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

aks681 committed:
Addressed previous review comments

aks681 committed:
modified answer outcomes for questions

aks681 committed:
Completed question editor directive frontend

aks681 committed:
fixed errors

aks681 committed:
addressed review comments

aks681 committed:
Fixed merge conflict

aks681 committed:
fixed errors

aks681 committed:
Added create question service

aks681 committed:
made review changes

aks681 committed:
Fixed e2e test

nithusha21 committed:
merge backend PR

aks681 committed:
made review changes

nithusha21 committed:
In progress

nithusha21 committed:
Merge remote-tracking branch 'aks-fork/question-services' into questions-interface

nithusha21 committed:
completed view for reviewers to view suggested questions

nithusha21 committed:
fix merge conflicts

nithusha21 committed:
bug fixes post merge

nithusha21 committed:
fix linting issues

nithusha21 committed:
more linting fixes

nithusha21 committed:
fix frontend test

seanlip commented on core/controllers/feedback.py:
```diff
@@ -64,6 +64,20 @@ def post(self, exploration_id):
 
 
+class ThreadListHandlerForTopics(base.BaseHandler):
+    """Handles listing of feedback threads linked to topics."""
```
feedback threads --> suggestion threads perhaps

nithusha21 replied:
Done

seanlip commented on core/controllers/feedback.py:
```diff
@@ -64,6 +64,20 @@ def post(self, exploration_id):
+    """Handles listing of feedback threads linked to topics."""
+    @acl_decorators.can_edit_topic
+    def get(self, topic_id):
+        # TODO (nithesh): Add non-suggestion feedback threads when required in
```
I think you can drop this TODO. If it's required we'll naturally add it.

nithusha21 replied:
Done

seanlip commented on core/controllers/suggestion.py:
```diff
@@ -94,6 +93,7 @@ def put(self, target_id, suggestion_id):
 class SuggestionToTopicActionHandler(base.BaseHandler):
     """Handles actions performed on suggestions to topics."""
 
+
```
Why add this line?

nithusha21 replied:
Removed

seanlip commented on core/controllers/suggestion.py:
```diff
@@ -119,7 +119,16 @@ def put(self, target_id, suggestion_id):
         suggestion = suggestion_services.get_suggestion_by_id(suggestion_id)
+
         if action == suggestion_models.ACTION_TYPE_ACCEPT:
+            if (
```
This is icky. Why don't you do it in the frontend before the suggestion comes in?

nithusha21 replied:
So we only pass the suggestion_id to the backend, and we retrieve the suggestion object and then accept. So we cannot populate the field in the frontend. I actually removed the function I had written, and updated suggestion.change.skill_id. This will save us one write operation to the DB. I think this is better.

seanlip replied:
OK, thanks for the explanation. I think this makes sense.

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -166,6 +166,22 @@ def _update_suggestion(suggestion):
     suggestion_model.put()
 
 
+def update_skill_id_for_question_suggestion(suggestion, skill_id):
```
See above; hopefully we can remove this.

Also if you keep this function, it needs tests.

nithusha21 replied:
I removed the function

seanlip commented on core/templates/dev/head/domain/suggestion/QuestionSuggestionObjectFactory.js:
```diff
@@ -0,0 +1,53 @@
+   suggestion domain objects.
+ */
+
+oppia.factory('QuestionSuggestionObjectFactory', ['QuestionObjectFactory',
```
Break after [

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/domain/suggestion/QuestionSuggestionObjectFactory.js:
```diff
@@ -0,0 +1,53 @@
+      this.skillId = skillId;
+    };
+
+    Suggestion.prototype.threadId = function() {
```
getThreadId. I hope this is fixed elsewhere too.

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -47,17 +47,19 @@ oppia.constant('HUMAN_READABLE_SUBSCRIPTION_SORT_BY_KEYS', {
   'HUMAN_READABLE_SUBSCRIPTION_SORT_BY_KEYS',
   function(
-      $scope, $rootScope, $http, $window, DateTimeFormatService,
+      $scope, $rootScope, $http, $uibModal, $window, DateTimeFormatService,
```
Line breaks here should match those in the stringified list.

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -189,6 +191,67 @@ oppia.controller('CreatorDashboard', [
+                $scope.questionStateData = $scope.question.getStateData();
+                $scope.topicSummaries = topicSummaries;
+
+                $scope.isNotValidQuestion = function() {
```
Flip this -- call it something like $scope.isQuestionValid. Clearer.

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -189,6 +191,67 @@ oppia.controller('CreatorDashboard', [
+                topicVersion = topicSummaries[i].version;
+                break;
+              }
+            }
```
And if none match, throw error

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -189,6 +191,67 @@ oppia.controller('CreatorDashboard', [
+                };
+
+                $scope.createQuestion = function() {
+                  $uibModalInstance.close({
```
Perhaps you should validate here that topicId is not null and that the question is valid.

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -189,6 +191,67 @@ oppia.controller('CreatorDashboard', [
+                question_dict: result.question.toBackendDict(true),
+                skill_id: null
+              },
+              description: 'question title'
```
Is this supposed to be hardcoded?

nithusha21 replied:
missed this comment!

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -189,6 +191,67 @@ oppia.controller('CreatorDashboard', [
+              target_version_at_submission: topicVersion,
+              change: {
+                cmd: 'create_new_fully_specified_question',
+                question_dict: result.question.toBackendDict(true),
```
@aks681 -- I looked at this in the codebase and it looks like the schemaVersion arg in toBackendDict() is unused. Can we remove it?

nithusha21 replied:
@aks681 ping!

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -189,6 +191,67 @@ oppia.controller('CreatorDashboard', [
+              description: 'question title'
+            });
+          }, function() {
+            $log('Error while submitting question');
```
$log.error

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/creator_dashboard/create_question_modal_directive.html:
```diff
@@ -0,0 +1,15 @@
+</div>
+
+<div class="modal-body">
+  <question-editor question-id="null" question-state-data="questionStateData" misconceptions="[]" can-edit-question="true">
```
Passing in a list like this (to "misconceptions") seems a bit fishy to me, since lists are not primitive values in JS. Perhaps you should declare a JS variable and bind to that instead.

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -42,6 +44,10 @@ oppia.directive('questionsTab', [
             $scope.questionSummaries =
               TopicEditorStateService.getQuestionSummaries();
+            $scope.questionSuggestionThreads = [];
+            $scope.getSuggestedQuestions();
```
Um, so, you get the suggested questions, but you don't assign them to anything?

Maybe better to write $scope.loadSuggestedQuestionsAsync() instead.

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -42,6 +44,10 @@ oppia.directive('questionsTab', [
               TopicEditorStateService.getQuestionSummaries();
+            $scope.questionSuggestionThreads = [];
+            $scope.getSuggestedQuestions();
+            $scope.questionUnderReview = null;
```
$scope.activeQuestion or $scope.questionCurrentlyBeingReviewed perhaps. In some sense all these questions are currently "under review".

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -52,7 +58,8 @@ oppia.directive('questionsTab', [
             EditableQuestionBackendApiService.createQuestion(
-              $scope.skillId, $scope.question.toBackendDict(true)
+              $scope.skillId, $scope.question.toBackendDict(
+                true)
```
Why this change? If you can fit an arg entirely on one line, do so.

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -118,6 +125,117 @@ oppia.directive('questionsTab', [
 
+          $scope.getSuggestedQuestions = function() {
+            $scope.questionSuggestionThreads = [];
+            $http.get('/generalsuggestionlisthandler', {
```
Look into $q.all(). Don't fire one after the other, do them together.

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -118,6 +125,117 @@ oppia.directive('questionsTab', [
+            });
+          };
+
+          $scope.setQuestionUnderReview = function(questionSuggestionThread) {
```
setActiveQuestion, setActiveSuggestionId, etc.

What happens to suggestionReviewMessage?

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -118,6 +125,117 @@ oppia.directive('questionsTab', [
+                  };
+
+                  $scope.done = function() {
+                    $uibModalInstance.close($scope.selectedSkillId);
```
So what actually happens to the selected skill ID after the modal is closed? AFAICT it isn't being used anywhere...

nithusha21 replied:
This was my bad. I fixed it now.

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -118,6 +125,117 @@ oppia.directive('questionsTab', [
+
+          $scope.acceptQuestion = function(suggestionId) {
+            $http.put(
+              '/generalsuggestionactionhandler/topic/' + $scope.topic.getId() +
```
Use UrlInterpolationService

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -118,6 +125,117 @@ oppia.directive('questionsTab', [
+
+          $scope.rejectQuestion = function(suggestionId) {
+            $http.put(
+              '/generalsuggestionactionhandler/topic/' + $scope.topic.getId() +
```
UrlInterpolationService

nithusha21 replied:
Done

seanlip commented on feconf.py:
```diff
@@ -324,7 +325,7 @@ def get_empty_ratings():
 # SYSTEM_EMAIL_ADDRESS is not that of an app owner, email messages from this
 # address cannot be sent. If True then emails can be sent to any user.
-CAN_SEND_EMAILS = False
+CAN_SEND_EMAILS = True
```
Don't flip this here. (We do turn it on in prod.)

nithusha21 replied:
Oops, done

seanlip reviewed (commented):
On the whole, structure looks pretty good! Did a full pass.

nithusha21 committed:
preliminary bug fixes

nithusha21 committed:
linting fixes

nithusha21 committed:
fix bug where user could not enter solution

nithusha21 committed:
fix loading of question editor directive using question state data

nithusha21 committed:
linting fix

nithusha21 commented:
@seanlip @anmolshkl, the 2 major frontend issues have been fixed (Thanks a lot @aks681).

anmolshkl commented:
Hi @nithusha21, apart from the bugs that we had already discussed I noticed a few other things: 

- Something is up with questions editor UI. When I try to create a suggestion for a question with multiple choice interaction, I'm not able to select the intended option from the drop-down menu. If I click on the 2nd choice, it selects 4th. Could you please try to replicate this issue and confirm if you can also see it?
- The modal for question suggestion should have `backdrop: 'static', keyboard: false` so as to prevent accidental closure of the form.
- The validations and warnings in the questions suggestion UI (creator dashboard) are not as strong as the actual question editor UI. For example, I can add a question without a hint while the question editor on the topic dashboard doesn't allow me to do that. Also, the editor doesn't tell me that I need to add a hint before submitting the question. One has to figure that out themselves.
- I'm unable to approve a suggestion. There seems to be some issue with linking question with the skill. In the UI, I clicked on `Select skill`, then clicked on approve and got greeted by the following message
`Expected change to contain skill_id`
- The review message box is too large, and, for a minute I didn't even notice it until I clicked on "Reject" and it said, "Reject message cannot be empty". But then I added a review message, and the validation still kept saying "Review message cannot be empty".
- I'm unable to reproduce the following bug but it pops up at random. When I refresh the page, sometimes the number of suggestions double up. For example, I created two questions but I can see two instances of each: 
![question-ui-2](https://user-images.githubusercontent.com/6338799/43999024-79ae86a8-9e20-11e8-9fc5-2e12158d0957.png)
It seems that this only happens on page refresh and if I navigate to the question editor from the dashboard again, then this issue disappears.

nithusha21 committed:
WIP

nithusha21 committed:
WIP

nithusha21 committed:
review changes, and bug fixes

nithusha21 committed:
linting fixes

nithusha21 commented:
> Something is up with questions editor UI. When I try to create a suggestion for a question with multiple choice interaction, I'm not able to select the intended option from the drop-down menu. If I click on the 2nd choice, it selects 4th. Could you please try to replicate this issue and confirm if you can also see it?

I am not able to reproduce this. I made 4 options [a, b, c, d] and selected b and it selected correctly.

> The modal for question suggestion should have backdrop: 'static', keyboard: false so as to prevent accidental closure of the form.

Done thanks!

> The validations and warnings in the questions suggestion UI (creator dashboard) are not as strong as the actual question editor UI. For example, I can add a question without a hint while the question editor on the topic dashboard doesn't allow me to do that. Also, the editor doesn't tell me that I need to add a hint before submitting the question. One has to figure that out themselves.

The submit button now stays disabled. Maybe an error message above the submit button as to what is the issue might be better I think. Will try to figure that out.


>I'm unable to approve a suggestion. There seems to be some issue with linking question with the skill. In the UI, I clicked on Select skill, then clicked on approve and got greeted by the following message
Expected change to contain skill_id

Yes, this was a bug. It is fixed now!

>The review message box is too large, and, for a minute I didn't even notice it until I clicked on "Reject" and it said, "Reject message cannot be empty". But then I added a review message, and the validation still kept saying "Review message cannot be empty".

The review message wasn't getting stored properly. I have fixed this issue now. I'll add some styling for the text box so that it looks better.

>I'm unable to reproduce the following bug but it pops up at random. When I refresh the page, sometimes the number of suggestions double up. For example, I created two questions but I can see two instances of each:
question-ui-2
It seems that this only happens on page refresh and if I navigate to the question editor from the dashboard again, then this issue disappears.

This is a race condition that is being caused by the _initTab function. This function is called twice. Sometimes however, it is called in quick succession, and though I empty the list on the top of the function that queries for suggestions, data from both these calls are stored. It doesn't occur on navigation as only one call to _initTab is being made then. 

I moved the function call outside the init to remove this error. There is one race condition I am facing because $scope.topic is not getting initialized by the time the suggestion query function is called. (Happens occasionally, so it feels like some race condition). I'll try to look into this. So if you get an error saying "interpolateURL takes only string parameters" or similar, the error *might* resolve on refreshing the page. I am investigating the issue.

nithusha21 committed:
minor fix

nithusha21 committed:
added informative error messages

nithusha21 committed:
linting fixes

nithusha21 commented:
>The validations and warnings in the questions suggestion UI (creator dashboard) are not as strong as the actual question editor UI. For example, I can add a question without a hint while the question editor on the topic dashboard doesn't allow me to do that. Also, the editor doesn't tell me that I need to add a hint before submitting the question. One has to figure that out themselves.

I have added an error message beside the submit button that mentions what else is needed before the question can  be submitted. I think this will help you know that you have to add a hint, and a solution (if applicable).

seanlip commented on core/controllers/creator_dashboard.py:
```diff
@@ -110,13 +115,38 @@ def get(self):
 class CreatorDashboardPage(base.BaseHandler):
     """Page showing the user's creator dashboard."""
 
+    EDITOR_PAGE_DEPENDENCY_IDS = ['codemirror']
```
This is misleading: call it ADDITIONAL_DEPENDENCY_IDS or similar. This isn't the editor page.

nithusha21 replied:
Done

seanlip commented on core/controllers/topics_and_skills_dashboard_test.py:
```diff
@@ -58,7 +58,7 @@ def test_get(self):
                 '%s' % feconf.TOPICS_AND_SKILLS_DASHBOARD_DATA_URL,
                 expect_errors=True)
-            self.assertEqual(response.status_int, 401)
+            self.assertEqual(response.status_int, 200)
```
Why is an "expect_errors=True" response giving a status code of 200?

More generally, why did this test change?

nithusha21 replied:
When a user tries to create a question, we need to list out the topics. But that user does not have permissions to get the topics list data. I had asked @aks681 what I should do, and he suggested to change the permissions of the get handler for this controller. I had made it open access, as any user should be able to create a question suggestion and we need to retrieve topic list for creation. 

As for the first question, I missed out on removing expect_errors=True. Will do!

seanlip replied:
This should not change. See above.

nithusha21 replied:
Done

seanlip commented on core/domain/question_domain.py:
```diff
@@ -82,7 +82,9 @@ def __init__(self, change_dict):
         if self.cmd == CMD_CREATE_NEW_FULLY_SPECIFIED_QUESTION:
             self.question_dict = change_dict['question_dict']
-            self.skill_id = change_dict['skill_id']
+            self.skill_id = None
```
Shouldn't we be passing "skill_id: null" explicitly in the change dict for this case? Each type of change_dict should have a fully-determined set of keys.

nithusha21 replied:
Ok will do!

nithusha21 replied:
Done. At creation, I am passing skill_id: None, in the change_dict.

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -408,13 +408,13 @@ def validate(self):
         before accepting the suggestion.
         """
+
+        if not self.change.skill_id:
```
if self.change.skill_id is None

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/creator_dashboard/create_question_modal_directive.html:
```diff
@@ -0,0 +1,15 @@
+</div>
+
+<div class="modal-body">
+  <question-editor question-id="null" question-state-data="questionStateData" misconceptions="[]" can-edit-question="true">
```
The previous comment I made on this was not addressed, although it was marked "Done". Don't pass lists directly as attr values.

nithusha21 replied:
I think I addressed it in another file and missed it here (Sorry for that). I had used lists in multiple places. Will correct it here.

nithusha21 replied:
Done

seanlip reviewed (commented):
This generally looks good. Just minor comments. Thanks!

anmolshkl commented:
Hi @nithusha21, please find my comments below - 
- Approving with empty and non-empty message gives "Expected change to contain skill_id". So, I'm still unable to approve any suggestion.
- Approve/Reject buttons do not get hidden once the suggestion is accepted or rejected (I was only able to reject suggestion). We should see something similar to common suggestion UI. Also, the third column in the table on the review page says "review" instead of rejected/accepted.
- Where can the suggestor see his suggestion status?
- The modal to create question should not show "End Exploration, "Continue" interactions.
- Also, there are minor issues with some of the interactions (I didn't test all of them). For example, on selecting Text Input interaction, the modal gives a warning - the correct solution does not lead to a new card. That doesn't make sense here.

nithusha21 commented:
> Approving with empty and non-empty message gives "Expected change to contain skill_id". So, I'm still unable to approve any suggestion.

Discussed on chat, we need to select skill using the "select skill" button before accepting.

> Approve/Reject buttons do not get hidden once the suggestion is accepted or rejected. We should see something similar to common sggestion UI. Also, the third column in the table on the review page says "review" instead of rejected/accepted.

Hmmm. On page refresh, everything works ok. Once status is not "review" the table rows are not clickable. Without refreshing, the data is not updated. I think I should force refresh the page on either accepting or rejecting a question.

> Where can the suggestor see his suggestion status?

At the moment, there isn't a view. I'll incorporate it with the common suggestion view to keep consistency. Can I take this up as a later task?

>The modal to create question should not show "End Exploration, "Continue" interactions.
Also, there are minor issues with some of the interactions (I didn't test all of them). For example, on selecting Text Input interaction, the modal gives a warning - the correct solution does not lead to a new card. That doesn't make sense here

Akshay is planning to restrict interactions in questions to only text input, multiple choice and numeric input (I think, correct me if I'm wrong!). So this will be resolved then. Also for your second part, did you first try an "end exploration" or "continue" and then delete it and try text input? That caused me some problems if I remember correctly, but as end exploration and continue are not going to be valid choices, this error will not occur. (However, if you managed to get the error with just text input, could you please elaborate on some repro steps? As I haven't faced this error when I did not use continue and end exploration interactions.

nithusha21 committed:
review changes

nithusha21 committed:
remove unused dependency

seanlip commented on core/domain/question_domain.py:
```diff
@@ -82,9 +82,7 @@ def __init__(self, change_dict):
-            self.skill_id = None
-            if 'skill_id' in change_dict:
-                self.skill_id = change_dict['skill_id']
+            self.skill_id = change_dict['skill_id']
```
Not a bad idea to add a comment above this line saying "Note that change_dict['skill_id'] may be None if this change is being done in the context of a suggestion."

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -24,13 +24,13 @@ oppia.directive('questionsTab', [
         '/pages/topic_editor/questions/questions_tab_directive.html'),
       controller: [
-        '$scope', '$http', '$q', '$uibModal', 'AlertsService',
+        '$scope', '$http', '$q', '$uibModal', '$window', 'AlertsService',
```
Why add $window? Doesn't seem to be used...

anmolshkl replied:
@nithusha21 this is causing an error preventing the creator dashboard to fully load - 
_Cause: undefined
$window is not defined
ReferenceError: $window is not defined_

nithusha21 replied:
Done

nithusha21 replied:
Actually I added it as I wanted to refresh the page after accept/reject. In that commit I hadn't made calls to the refresh function. The refresh is called now.

seanlip reviewed (commented):
Looks good! Just some minor comments on the diff.

seanlip commented on core/controllers/topics_and_skills_dashboard.py:
```diff
@@ -48,7 +48,9 @@ class TopicsAndSkillsDashboardPageDataHandler(base.BaseHandler):
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
 
-    @acl_decorators.can_access_topics_and_skills_dashboard
+    # As anyone should be able to see the topics and skills list when creating
```
I'm reading this function more carefully. It doesn't look like it returns the skills list. It only returns untriaged/unpublished skills. It also returns a lot of data that is not relevant to what you need in the creator dashboard.

More generally, I think it's not a good idea to couple this handler with the creator dashboard. Could we instead just leave this data handler as-is, and create a new handler for the creator dashboard that returns specifically the info that is required? Otherwise it's not clear to developers working on this handler that the handler serves a completely different unrelated page as well.

anmolshkl replied:
@nithusha21 have we addressed this?

nithusha21 replied:
Done

nithusha21 replied:
I moved this data retrieval to the creator dashboard data handler itself (don't know why I didn't think of that earlier). It fits into the pattern used to retrieve exploration, and collection data in the creator dashboard.

seanlip commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -189,6 +191,86 @@ oppia.controller('CreatorDashboard', [
+            question_dict: result.question.toBackendDict(true),
+            skill_id: null
+          },
+          description: 'question title'
```
What's this "description"? For the exploration feedback threads it is:

        description: str. Learner-provided description of suggestion changes.

So, if it's like a commit message, perhaps we should have something a bit more clear? Like, "Update the question title"? Or "Create new question"? Or "Initialize question with title XXX"?

Basically, I don't understand the value for this field. Please explain what it should do and pick an appropriate value for it.

nithusha21 replied:
Initially I thought in the questions tab, I will show the description and the status (Similar to the feedback tab view for explorations, where the suggestion descriptions are shown to you). However, for consistency with the other questions being displayed on that page, I had used question state data in that view. So this is actually not being used! I'll remove it. 

Sorry for the confusion.

nithusha21 replied:
Done

seanlip commented on core/controllers/feedback.py:
```diff
@@ -64,6 +64,18 @@ def post(self, exploration_id):
         self.render_json(self.values)
 
 
+class ThreadListHandlerForTopics(base.BaseHandler):
```
Is this functionality live? If not should it be behind a flag?

anmolshkl replied:
@nithusha21 ping?

nithusha21 replied:
Done. It is hidden behind both ENABLE_GENERALIZED_FEEDBACK_THREADS and ENABLE_NEW_STRUCTURES

seanlip commented on core/templates/dev/head/pages/creator_dashboard/creator_dashboard.html:
```diff
@@ -56,6 +61,13 @@ <h1 class="stat-value-without-rating" ng-hide="dashboardStats.num_ratings || rel
         </div>
       </md-card>
     </div>
+    <div class="row">
```
Is all this new functionality supposed to go fully live in the upcoming release? If so, I'd be surprised, since Questions functionality is not generally released yet.

If it's not supposed to go fully live yet, shouldn't this stuff be behind an ng-if condition?

nithusha21 replied:
No this isn't ready for release (Will get released along with the new structures project I guess). Yep, I missed adding a flag, will do so. Shall I create a new flag, (as otherwise it will be a combination of ENABLE_GENERALIZED_FEEDBACK_THREADS and ENABLE_NEW_STRUCTURES).

seanlip replied:
Oh dear. That's a major omission, glad it got caught.

Don't create a new flag. Use a combination of both existing flags.

anmolshkl replied:
@nithusha21 ping?

nithusha21 replied:
Done

seanlip commented on core/templates/dev/head/pages/topic_editor/questions/QuestionsTabDirective.js:
```diff
@@ -42,6 +45,10 @@ oppia.directive('questionsTab', [
             $scope.canEditQuestion = $scope.topicRights.canEditTopic();
             $scope.questionSummaries =
               TopicEditorStateService.getQuestionSummaries();
+            $scope.emptyMisconceptionsList = [];
```
I get that this starts off empty. But will it be empty forever? Wouldn't people who suggest questions be adding misconceptions to them too?

anmolshkl replied:
@nithusha21 also, there doesn't seem to be a provision to add misconception in the UI.

nithusha21 replied:
For the initial version of this feature, I have allowed the question suggestion creator to select a topic, and then the reviewer must link the question to a skill. To tag answers with misconceptions, we need to know which skill is (going to be) linked with the question.  However, at the point where the reviewer links the question to a skill, there is no provision to update the question (yet). So I think once we create a way to update suggestions once they have been submitted, I can add this feature. But I have excluded it in the first version of the questions suggestion feature.

anmolshkl replied:
So, the onus is on the reviewer to link a misconception? NB: lets do this post GSoC, please file an issue for this.

@seanlip On a side note, I'm still not very clear with the concept of "misconception" and how it is linked to question or a skill. As a creator, I feel pretty confused about these new terminologies (i.e. new structures) and I'm slightly concerned that creating lessons has a steep learning curve now (or maybe I just don't understand it). Having tried creating questions, topics and skills extensively, I have a few things to discuss with you post GSoC.

nithusha21 replied:
Yes, I'll file an issue for this. From the current UX, the reviewer links a skill (and only then misconceptions are "defined", as misconceptions link to skills). So I guess it would be the responsibility of the reviewer finally.

anmolshkl commented:
@nithusha21 I'm unable to suggest questions due to a bug in the creator dashboard. Also, there still seems to be an unresolved bug on the create page - `Exception: Not allowed to query on field list_type`.

seanlip commented:
Please note merge conflicts.

nithusha21 committed:
partial review changes

nithusha21 committed:
fix merge conflicts

nithusha21 committed:
minor fixes

nithusha21 committed:
remove hardcoded description value

nithusha21 committed:
fixes

anmolshkl commented:
@nithusha21 I have taken a look. The UI works fine. Although there are many outstanding issues, but the very basic functionalities work alright. Lets file a separate issue for all the remaining functionalities (eg, displaying suggestions in common suggestion view for creator and suggestor etc).

nithusha21 committed:
fixes, remove invalid scripts, and remove comment

nithusha21 committed:
linting fix

nithusha21 commented:
> @nithusha21 I'm unable to suggest questions due to a bug in the creator dashboard. Also, there still seems to be an unresolved bug on the create page - Exception: Not allowed to query on field list_type.

This has been fixed on merging the PR with general bug fixes for the feedback framework. The bug in the creator dashboard has been resolved by importing $window again (was accidentally removed!)

anmolshkl commented on core/templates/dev/head/pages/state_editor/StateResponsesDirective.js:
```diff
@@ -576,6 +577,11 @@ oppia.directive('stateResponses', [
             return outcome && (outcome.dest === activeStateName);
           };
+
+
```
Double blank lines?

anmolshkl reviewed (approved):
@nithusha21 lgtm!

nithusha21 committed:
remove extra new line

seanlip commented:
Is it safe to merge this into develop?

nithusha21 committed:
fix merge conflicts

nithusha21 committed:
linting fixes

anmolshkl commented:
@seanlip yep, once the travis build gets completed, we can merge this.

bansalnitish commented:
@seanlip @anmolshkl all green here! Merging this one. 

Thanks @nithusha21

vojtechjelinek commented on core/templates/dev/head/pages/creator_dashboard/creator_dashboard.html:
```diff
@@ -838,7 +867,207 @@ <h2>Suggestions to review</h2>
 
 {% block footer_js %}
   {{ super() }}
+  <script src="/templates/dev/head/mathjaxConfig.js"></script>
```
Hi @nithusha21, do you remember if when you were adding this script-import if you copied from somewhere or if you added only script-imports that were really needed for this page?

nithusha21 replied:
This page displays the state editor(for creation of questions). For which I think we need this import. It was technically a copy over from the other questions editor.

vojtechjelinek replied:
Sorry, I phrased my question wrong, I was talking about all the script-imports in this file because some of them seem unused.

nithusha21 replied:
Oh I see, I am not sure, but I think it was a direct copy paste from the exploration editor or the question editor page. They might be unused (I don't recollect checking each one of them).

vojtechjelinek replied:
Ok, thanks for the info.
